### PR TITLE
Load annotations on demand

### PR DESCRIPTION
**Pull request by dominikl**
Cherrypicked from #4654 (closed the original PR).

Up to now whenever an object is selected all annotations for this object are loaded. As we now have the different taskpanes on the right hand side metadata panel, which are closed by default, this is not necessary any more. With this PR only the specific annotations are loaded on demand, i.e when a taskpane is expanded. While the annotations are loaded from the server, the expanded taskpane displays a simple "Loading..." label.

I started this more as a kind of proof of concept, but it turned into a quite extensive PR. Which means there's a risk that it'll break something and needs a lot of testing (basically everything annotation related). Is it worth it? Taking into account, that this PR won't add a noticeable performance increase in most cases, only for data objects with a very large amount of annotations attached to them.

**Test**
Basically test right hand side metadata panel. Also worth testing with "other" annotations, e.g. download and import an image from JCB, they usually have some long and/or boolean annotations.

--exclude

### CODE REVIEW

**Comment by dominikl** [comment]
Could someone please give me a hint, what the problem with Travis, respectively the commit 
ea47864 might be? I'm trying to get Travis for #4654 green again. As I have no idea what causes the problem, I opened this PR, cherry-picking and building each individual commit to narrow it down a bit. Apparently it's already the second commit which is the problem. But still I have no idea why...

From the build log:
```
[nomemorytestng] 2016-09-21 16:02:03,649 ERROR [ome.services.db.SelfCorrectingDataSource] (oRetries()) Failed to acquire connection after retries=0
[nomemorytestng] ome.server.utests.SelfCorrectingDatabaseUnitTest$MySQLException: null
[nomemorytestng] 	at org.jmock.core.stub.ThrowStub.invoke(Unknown Source) ~[jmock-1.0.1.jar:na]
[nomemorytestng] 	at org.jmock.core.InvocationMocker.invoke(Unknown Source) ~[jmock-1.0.1.jar:na]
[nomemorytestng] 	at org.jmock.core.LIFOInvocationDispatcher.dispatch(Unknown Source) ~[jmock-1.0.1.jar:na]
[nomemorytestng] 	at org.jmock.core.AbstractDynamicMock.mockInvocation(Unknown Source) ~[jmock-1.0.1.jar:na]
[nomemorytestng] 	at org.jmock.core.CoreMock.invoke(Unknown Source) ~[jmock-1.0.1.jar:na]
[nomemorytestng] 	at com.sun.proxy.$Proxy70.getConnection(Unknown Source) ~[na:na]
[nomemorytestng] 	at org.springframework.jdbc.datasource.DelegatingDataSource.getConnection(DelegatingDataSource.java:83) ~[spring-jdbc-3.0.1.RELEASE.jar:3.0.1.RELEASE]
```
I have no clue, exasperating...

**Comment by mtbc** [comment]
I don't know about the frustrating Travis issue but maybe later you could run OMERO-DEV-merge-push with status: none, then OMERO-DEV-merge-build, so people can at least download the artifact and test?

**Comment by dominikl** [comment]
That's a good idea. I'll kick off a build with "status: none" tonight, to not interfere with the daily builds.

**Comment by mtbc** [comment]
I think once the morning one's deployed (i.e. now) it's just a case of giving the morning's Insight testers enough time to download the "kosher" artifact first for the other relevant PRs.

**Comment by dominikl** [comment]
Looks like the cause was an unintentional wrong import of `edu.emory.mathcs.backport.java.util.Collections` (once again...). At least I hope so. Will now cherry-pick the other commits.

**Comment by mtbc** [comment]
Maybe you need a local git commit hook. :smiley:

**Comment by dominikl** [comment]
I can't believe, that Travis is green, finally. Updated PR description and closed the original PR.

**Comment by jburel** [comment]
time to review it?

**Comment by dominikl** [comment]
Nearly, I just noticed on issue with MapAnnotations, just wanted to check again, if that's caused by this PR or something else. I'll add the PR to standup then.

**Comment by pwalczysko** [comment]
Bug (blocker): multiselect 3 images in central pane, observe error and crash. Fully repeatable.

```
java.lang.Exception: Abnormal termination due to an uncaught exception.
java.lang.NullPointerException
    at java.util.ArrayList.addAll(ArrayList.java:577)
    at org.openmicroscopy.shoola.agents.metadata.view.MetadataViewerComponent.loadAnnotationCount(MetadataViewerComponent.java:1348)
    at org.openmicroscopy.shoola.agents.metadata.editor.EditorModel.loadAnnotationCount(EditorModel.java:4653)
    at org.openmicroscopy.shoola.agents.metadata.editor.GeneralPaneUI.setRootObject(GeneralPaneUI.java:630)
    at org.openmicroscopy.shoola.agents.metadata.editor.EditorUI.setRootObject(EditorUI.java:363)
    at org.openmicroscopy.shoola.agents.metadata.editor.EditorComponent.setRootObject(EditorComponent.java:294)
    at org.openmicroscopy.shoola.agents.metadata.view.MetadataViewerModel.setRootObject(MetadataViewerModel.java:335)
    at org.openmicroscopy.shoola.agents.metadata.view.MetadataViewerComponent.setRootObject(MetadataViewerComponent.java:420)
    at org.openmicroscopy.shoola.agents.treeviewer.view.TreeViewerComponent.setSelectedNode(TreeViewerComponent.java:1497)
    at org.openmicroscopy.shoola.agents.treeviewer.view.TreeViewerComponent.setSelectedNodes(TreeViewerComponent.java:1424)
    at org.openmicroscopy.shoola.agents.treeviewer.view.TreeViewerControl.propertyChange(TreeViewerControl.java:1530)
    at org.openmicroscopy.shoola.util.ui.component.AbstractComponent.firePropertyChange(AbstractComponent.java:147)
    at org.openmicroscopy.shoola.agents.dataBrowser.view.DataBrowserComponent.setSelectedDisplays(DataBrowserComponent.java:363)
    at org.openmicroscopy.shoola.agents.dataBrowser.view.DataBrowserControl.propertyChange(DataBrowserControl.java:499)
    at org.openmicroscopy.shoola.util.ui.component.AbstractComponent.firePropertyChange(AbstractComponent.java:147)
    at org.openmicroscopy.shoola.agents.dataBrowser.browser.BrowserModel.setSelectedDisplays(BrowserModel.java:875)
    at org.openmicroscopy.shoola.agents.dataBrowser.browser.BrowserControl.mouseReleased(BrowserControl.java:635)
    at java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:290)
    at java.awt.Component.processMouseEvent(Component.java:6533)
    at javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
    at java.awt.Component.processEvent(Component.java:6298)
    at java.awt.Container.processEvent(Container.java:2236)
    at java.awt.Component.dispatchEventImpl(Component.java:4889)
    at java.awt.Container.dispatchEventImpl(Container.java:2294)
    at java.awt.Component.dispatchEvent(Component.java:4711)
    at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4888)
    at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4525)
    at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4466)
    at java.awt.Container.dispatchEventImpl(Container.java:2280)
    at java.awt.Window.dispatchEventImpl(Window.java:2746)
    at java.awt.Component.dispatchEvent(Component.java:4711)
    at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
    at java.awt.EventQueue.access$500(EventQueue.java:97)
    at java.awt.EventQueue$3.run(EventQueue.java:709)
    at java.awt.EventQueue$3.run(EventQueue.java:703)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
    at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
    at java.awt.EventQueue$4.run(EventQueue.java:731)
    at java.awt.EventQueue$4.run(EventQueue.java:729)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
    at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
    at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
    at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
    at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
    at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
    at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
    at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Abnormal termination due to an uncaught exception.
java.lang.NullPointerException
    at java.util.ArrayList.addAll(ArrayList.java:577)
    at org.openmicroscopy.shoola.agents.metadata.view.MetadataViewerComponent.loadAnnotationCount(MetadataViewerComponent.java:1348)
    at org.openmicroscopy.shoola.agents.metadata.editor.EditorModel.loadAnnotationCount(EditorModel.java:4653)
    at org.openmicroscopy.shoola.agents.metadata.editor.GeneralPaneUI.setRootObject(GeneralPaneUI.java:630)
    at org.openmicroscopy.shoola.agents.metadata.editor.EditorUI.setRootObject(EditorUI.java:363)
    at org.openmicroscopy.shoola.agents.metadata.editor.EditorComponent.setRootObject(EditorComponent.java:294)
    at org.openmicroscopy.shoola.agents.metadata.view.MetadataViewerModel.setRootObject(MetadataViewerModel.java:335)
    at org.openmicroscopy.shoola.agents.metadata.view.MetadataViewerComponent.setRootObject(MetadataViewerComponent.java:420)
    at org.openmicroscopy.shoola.agents.treeviewer.view.TreeViewerComponent.setSelectedNode(TreeViewerComponent.java:1497)
    at org.openmicroscopy.shoola.agents.treeviewer.view.TreeViewerComponent.setSelectedNodes(TreeViewerComponent.java:1424)
    at org.openmicroscopy.shoola.agents.treeviewer.view.TreeViewerControl.propertyChange(TreeViewerControl.java:1530)
    at org.openmicroscopy.shoola.util.ui.component.AbstractComponent.firePropertyChange(AbstractComponent.java:147)
    at org.openmicroscopy.shoola.agents.dataBrowser.view.DataBrowserComponent.setSelectedDisplays(DataBrowserComponent.java:363)
    at org.openmicroscopy.shoola.agents.dataBrowser.view.DataBrowserControl.propertyChange(DataBrowserControl.java:499)
    at org.openmicroscopy.shoola.util.ui.component.AbstractComponent.firePropertyChange(AbstractComponent.java:147)
    at org.openmicroscopy.shoola.agents.dataBrowser.browser.BrowserModel.setSelectedDisplays(BrowserModel.java:875)
    at org.openmicroscopy.shoola.agents.dataBrowser.browser.BrowserControl.mouseReleased(BrowserControl.java:635)
    at java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:290)
    at java.awt.Component.processMouseEvent(Component.java:6533)
    at javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
    at java.awt.Component.processEvent(Component.java:6298)
    at java.awt.Container.processEvent(Container.java:2236)
    at java.awt.Component.dispatchEventImpl(Component.java:4889)
    at java.awt.Container.dispatchEventImpl(Container.java:2294)
    at java.awt.Component.dispatchEvent(Component.java:4711)
    at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4888)
    at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4525)
    at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4466)
    at java.awt.Container.dispatchEventImpl(Container.java:2280)
    at java.awt.Window.dispatchEventImpl(Window.java:2746)
    at java.awt.Component.dispatchEvent(Component.java:4711)
    at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
    at java.awt.EventQueue.access$500(EventQueue.java:97)
    at java.awt.EventQueue$3.run(EventQueue.java:709)
    at java.awt.EventQueue$3.run(EventQueue.java:703)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
    at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
    at java.awt.EventQueue$4.run(EventQueue.java:731)
    at java.awt.EventQueue$4.run(EventQueue.java:729)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
    at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
    at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
    at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
    at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
    at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
    at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
    at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Exception in thread "AWT-EventQueue-0"

    at org.openmicroscopy.shoola.env.ui.UserNotifierImpl.showErrorDialog(UserNotifierImpl.java:189)
    at org.openmicroscopy.shoola.env.ui.UserNotifierImpl.notifyError(UserNotifierImpl.java:289)
    at org.openmicroscopy.shoola.env.AbnormalExitHandler.doTermination(AbnormalExitHandler.java:147)
    at org.openmicroscopy.shoola.env.AbnormalExitHandler.terminate(AbnormalExitHandler.java:85)
    at org.openmicroscopy.shoola.env.RootThreadGroup.uncaughtException(RootThreadGroup.java:69)
    at java.awt.EventDispatchThread.processException(EventDispatchThread.java:223)
    at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:215)
    at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
    at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
    at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
    at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
    at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

**Comment by pwalczysko** [comment]
Not sure why the tabs have `(-)` after their name when well images are selected, see for example Tags harmonica in RHP screenshot below

screen shot 2016-10-06 at 11 03 54

**Comment by pwalczysko** [comment]
When I manage to multiselect without crash (by multiselecting in the tree) I observe that the RHP is unusable with regard to Tags, Attachments, Ratings and Comments (you can write a comment but it will not appear after you hit OK). All these harmonicas are missing controls for annotation addition or removal.
screen shot 2016-10-06 at 11 07 23

**Comment by pwalczysko** [comment]
https://github.com/openmicroscopy/openmicroscopy/pull/4854#issuecomment-251917489 and https://github.com/openmicroscopy/openmicroscopy/pull/4854#issuecomment-251920606 looks like pretty serious, workflow destroying bugs to me already to justify stopping testing until they are fixed.
More granular testing can follow once this is done.

In summary, up till now 3 bugs found.

**Comment by dominikl** [comment]
All three bugs should be fixed now @pwalczysko .

**Comment by pwalczysko** [comment]
Select 3 images, add 3 tags onto them. then go and check the tags presence in RHP. On multiselect of 4 images, the tags tab shows nothing for 3 sec, then the tags come. 
Compare with latest, the wait is just less than a sec.

**Comment by dominikl** [comment]
The reason for that is, with latest, all metadata is loaded as soon as you click on an image/dataset/... So there should be actually no delay at all, when opening the tags tab. Whereas with this PR in, the tag metadata is loaded when you open the tags tab (same for other attachments, comments, etc.), that's why there's a delay. But it actually should not display nothing, there should be a "Loading" label while the data is retrieved. I have to check that.

**Comment by pwalczysko** [comment]
Even if there is no tag present, there should be the word "None" displayed in the opened harmonica. This is unfortunately not the case in some 30-40 percent of the images (with no tags). It appears that this behaviour is pretty random, but easily repeatable if you switch between images (have some 10 images to play with with no tags on them).

**Comment by dominikl** [comment]
Thanks @pwalczysko , I'll have to check that.

**Comment by pwalczysko** [comment]
If I select "Show added by others" (there are no annotations added by others on the image in question), then I still have in the Key-Value Pair harmonica following "ghost" Key and Value displayed:
screen shot 2016-10-07 at 10 32 33

**Comment by pwalczysko** [comment]
Show added by others ignores Ratings - these are still shown, although they were added by me.

**Comment by pwalczysko** [comment]
The counter on "Attachments" seems to be confused on an image, I added 2 attachments, it shows 3:
screen shot 2016-10-07 at 10 36 57

**Comment by pwalczysko** [comment]
The Show added by me dropdown has to be completely reviewed here. Except https://github.com/openmicroscopy/openmicroscopy/pull/4854#issuecomment-252195900 and https://github.com/openmicroscopy/openmicroscopy/pull/4854#issuecomment-252196440 I also see that Show added by others shows actually the same thing as Show added by me. I will ignore further bugs in this section suggesting @dominikl will review this small menu behaviour in this PR completely.

**Comment by pwalczysko** [comment]
The counts are wrong in multi-selection cases. Tags seem to be counted per-image and added together, albeit these are the same tags (e.g. have 5 tags on 3 images and multiselect the 3 images, the counter shows 15 tags, the tab itself shows correctly just 5).
In case of Comments I do not see any logic there, see screenshot below.

screen shot 2016-10-07 at 10 46 58

**Comment by pwalczysko** [comment]
Bug: Jumping to the right.
- select 3 images
- expand Rating
- add Rating and observe that as the new Rating is saved, it "jumps" a bit to the right
- change the added Rating by clicking on it again, adding or removing a start
- observe again a jump of the stars to the right in the tab
- repeat until the RHP becomes so long that the controls on the right are not visible

**Comment by dominikl** [comment]
Thanks for testing. There are lot of bugs to fix. I'll exclude the PR for now, until that's done.

**Comment by dominikl** [comment]
Closing again... this is just too unstable.
